src/zenml/plugins/plugin_flavor_registry.py

```diff
@@ -0,0 +1,364 @@
+from zenml.plugins.base_plugin_flavor import BasePlugin, BasePluginFlavor
+
+logger = get_logger(__name__)
+if TYPE_CHECKING:
+    from zenml.models.v2.base.base_plugin_flavor import (
+        BasePluginFlavorResponse,
+    )
+
+
+class RegistryEntry(BaseModel):
+    """Registry Entry Class for the Plugin Registry."""
+
+    flavor_class: Type[BasePluginFlavor]
+    plugin_instance: Optional[BasePlugin]
+
+    class Config:
+        """Pydantic configuration class."""
+
+        arbitrary_types_allowed = True
+
+
+class PluginFlavorRegistry:
+    """Registry for plugin flavors."""
+
+    def __init__(self) -> None:
+        """Initialize the event flavor registry."""
+        self.plugin_flavors: Dict[
+            PluginType, Dict[PluginSubType, Dict[str, RegistryEntry]]
+        ] = {}
+        self.register_plugin_flavors()
+
+    @property
+    def _types(self) -> List[PluginType]:
+        """Returns all available types.
+
+        Returns:
+            List of all available plugin types.
+        """
+        return list(self.plugin_flavors.keys())
+
+    def list_subtypes_within_type(
+        self, _type: PluginType
+    ) -> List[PluginSubType]:
+        """Returns all available subtypes for a given type.
+
+        Args:
+            _type: The type of plugin
+
+        Returns:
+            A list of available plugin subtypes for this plugin type.
+        """
+        return list(self.plugin_flavors[_type].keys())
+
+    def _flavor_entries(
+        self, _type: PluginType, subtype: PluginSubType
+    ) -> Dict[str, RegistryEntry]:
+        """Get a list of all subtypes for a specific flavor and type.
+
+        Args:
+            _type: The type of Plugin
+            subtype: The subtype of the plugin
+
+        Returns:
+            Dict of Registry entries sorted by flavor name.
+        """
+        if (
+            _type in self.plugin_flavors
+            and subtype in self.plugin_flavors[_type]
+        ):
+            return self.plugin_flavors[_type][subtype]
+        else:
+            return {}
+
+    def list_available_flavor_names_for_type_and_subtype(
+        self,
+        _type: PluginType,
+        subtype: PluginSubType,
+        page: int,
+        size: int,
+    ) -> List[str]:
+        """Get a list of all subtypes for a specific flavor and type.
+
+        Args:
+            _type: The type of Plugin
+            subtype: The subtype of the plugin
+            page: For optional pagination
+            size: For optional pagination
+
+        Returns:
+            List of flavor names for the given type/subtype combination.
+
+        Raises:
+            ValueError: If the page or size are invalid.
+        """
+        flavor_names = list(
+            self._flavor_entries(_type=_type, subtype=subtype).keys()
+        )
+        if page > 0 and size > 0:
+            start = page * size
+            end = start + size
+            return flavor_names[start:end]
+        else:
```

Comment:
is there a scenario where this can go beyond lenght of `flavor_names`?

Reply:
good catch, this function isn't even used anymore, it was replaced by `list_available_flavor_responses_for_type_and_subtype`. I checked and fixed that one instead.